File: daily_action_cycle_mobile_ui/lib/core/entities/day.dart
import 'package:equatable/equatable.dart';
import 'activity.dart';
import 'action_template.dart';

class Day extends Equatable {
  final String id;
  final DateTime date;
  final String actionTemplateId;
  final ActionTemplate? actionTemplate;
  final List<Activity> tasks;

  Day({
    required this.id,
    required this.date,
    required this.actionTemplateId,
    this.actionTemplate,
    required this.tasks,
  });

  @override
  List<Object?> get props =>
      [id, date, actionTemplateId, actionTemplate, tasks];
}

File: DailyActionCycle.WebAPI/DailyActionCycle.WebUI/Entities/Day.cs
namespace DailyActionCycle.WebUI.Entities;

public sealed record class Day
{
    public Day(DateOnly date)
    {
        Id = Guid.NewGuid();
        Date = date;
    }

    public Guid Id { get; init; }

    public DateOnly Date { get; init; }

    public Guid? ActionTemplateId { get; set; }
    public ActionTemplate? ActionTemplate { get; set; }

    public List<Activity> Activities { get; set; } = [];


    public IEnumerable<Activity> IncompletedActivities() => Activities
        .Where(activity => !activity.IsCompleted)
        .AsEnumerable();

    public IEnumerable<Activity> CompletedActivities() => Activities
        .Where(activity => activity.IsCompleted)
        .AsEnumerable();

    public void AddActivity(Activity activity)
    {
        Activities.Add(activity);
    }

    public void RemoveActivity(Activity activity)
    {
        Activities.Remove(activity);
    }

    public void CompleteActivity(Activity activity)
    {
        Activities.Find(t => t.Id == activity.Id)?.Complete();
    }
}
